File: android/app/src/main/java/com/openrung/vpn/EngineEventDispatcher.kt
package com.openrung.vpn

import io.nekohasekai.libbox.OpenRungEngineListener
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull

/** A B1 callback copied onto the Android service queue before any native state changes. */
internal data class EngineEvent(val sequence: Long, val kind: String, val payload: JsonObject)

/**
 * Process-lifetime gomobile listener. [post] must enqueue asynchronously on the service queue;
 * neither parsing nor a receiver may run inline on the Go callback goroutine (it holds Go locks).
 *
 * Attach/detach run on that same queue. Before attaching a successor, the host must join the
 * previous engine Stop: this gate discards queued deliveries, not future callbacks from a run
 * the host left alive. Engine sequence numbers persist across service owners and sessions.
 */
internal class EngineEventDispatcher(
    private val post: (() -> Unit) -> Unit,
    private val onInvalidEvent: (String) -> Unit,
) : OpenRungEngineListener {
    private class Owner(val receive: (EngineEvent) -> Unit)
    @Volatile private var owner: Owner? = null
    private var lastSequence = 0L // service queue only

    fun attach(receive: (EngineEvent) -> Unit) {
        owner = Owner(receive)
    }

    fun detach() {
        owner = null
    }

    override fun onEvent(eventJSON: String?) {
        val destination = owner ?: return
        post {
            if (owner !== destination) return@post
            val event = runCatching { decode(eventJSON) }.getOrElse {
                // Diagnostic excludes payloads, which may contain connection details.
                onInvalidEvent("invalid connectcore event envelope")
                return@post
            }
            if (event.sequence <= lastSequence) return@post
            lastSequence = event.sequence
            destination.receive(event)
        }
    }

    private fun decode(raw: String?): EngineEvent {
        requireNotNull(raw)
        val envelope = Json.parseToJsonElement(raw) as? JsonObject ?: error("object required")
        require(envelope["version"]?.jsonPrimitive?.intOrNull == 1)
        val sequence = envelope["sequence"]?.jsonPrimitive?.longOrNull ?: error("sequence required")
        require(sequence > 0)
        val kind = envelope["kind"]?.jsonPrimitive?.contentOrNull ?: error("kind required")
        require(kind in setOf("state", "notice", "log"))
        val payload = envelope["payload"] as? JsonObject ?: error("payload required")
        return EngineEvent(sequence, kind, payload)
    }
}

File: ios/Shared/EngineEventDispatcher.swift
import Foundation

struct EngineEvent {
    let sequence: UInt64
    let kind: String
    let payload: [String: Any]

    var status: ConnectionStatus? {
        guard kind == "state", let value = payload["Status"] as? String else { return nil }
        return ConnectionStatus(rawValue: value.lowercased())
    }
}

/// Go callbacks only enqueue. Attachment identity is captured BEFORE enqueueing,
/// so buffered events cannot update a replacement provider. Attach/detach and
/// delivery run on the host's serial queue; the capture itself is lock protected.
final class EngineEventDispatcher {
    private final class Attachment {
        let receive: (EngineEvent) -> Void
        init(_ receive: @escaping (EngineEvent) -> Void) { self.receive = receive }
    }
    private let lock = NSLock()
    private var attachment: Attachment?
    private var lastSequence: UInt64 = 0
    private let post: (@escaping () -> Void) -> Void
    private let invalid: () -> Void

    init(post: @escaping (@escaping () -> Void) -> Void, invalid: @escaping () -> Void = {}) {
        self.post = post
        self.invalid = invalid
    }

    func attach(_ receive: @escaping (EngineEvent) -> Void) {
        lock.lock()
        attachment = Attachment(receive)
        lastSequence = 0
        lock.unlock()
    }

    func detach() {
        lock.lock()
        attachment = nil
        lock.unlock()
    }

    func onEvent(_ raw: String?) {
        lock.lock()
        let destination = attachment
        lock.unlock()
        guard let destination else { return }
        post { [self] in
            lock.lock()
            let current = attachment === destination
            lock.unlock()
            guard current else { return }
            guard let raw, let data = raw.data(using: .utf8),
                  let envelope = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let version = envelope["version"] as? NSNumber, CFGetTypeID(version) != CFBooleanGetTypeID(), version == 1,
                  let number = envelope["sequence"] as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID(),
                  let sequence = UInt64(number.stringValue), sequence > 0,
                  let kind = envelope["kind"] as? String, !kind.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                  let payload = envelope["payload"] as? [String: Any] else { invalid(); return }
            guard sequence > lastSequence else { return }
            lastSequence = sequence
            guard ["state", "notice", "log"].contains(kind) else { return }
            destination.receive(EngineEvent(sequence: sequence, kind: kind, payload: payload))
        }
    }
}
